doc/sphinx/proofs/writing-proofs/proof-mode.rst

```diff
@@ -4,6 +4,87 @@
  Proof handling
 -------------------
 
+.. todo PR: my example is pretty bad--can you suggest a better one?
+   Also need to position/integrate this better
+
+Proof State
+-----------
+
+Proof state consists of one or more subgoals that need to be proven.
+Each subgoal has a :gdef:`conclusion` (the statement that is to be proven)
+and may have zero or more :gdef:`hypotheses <hypothesis>`,
+which are specific facts that can be used to proving the conclusion (in
+addition to definitions, proven theorems, etc.).
+
+Hypotheses appear above the line and the conclusion appears below the line.
+Both of them are types.  Hypotheses have names, which appear before the colon
+and may be referred to in the conclusion or other hypotheses.
+Some hypotheses are akin to specifying the type of a variable, for example `n : nat`.
+
+The term ":gdef:`goal`" may refer to an entire subgoal or to the conclusion
+of a subgoal, depending on the context.
+
+.. todo PR where does the term "hypothesis" come from?  It is not the
+   common definition used in science.  Does it come from elsewhere or
+   it Coq's invention to use it this way?  I could observe that it's
+   a misnomer
+
+The proof state when you begin proving a theorem shows
+the statement of the theorem below the line and no hypotheses above the
+line:
+
+.. coqtop:: none
+
+   Parameter P: nat -> Prop.
+
+.. coqtop:: out
+
+   Goal forall n m: nat, n > m -> P 1 /\ P 2.
+
+After applying a :term:`tactic` (shown below), we see hypotheses above the line.  The names of
+hypotheses (`n`, `m` and `H`) appear before a colon, followed by the type
```

Comment:
Note that what we usually call "hypothesis" are only the "H" (i.e. the variables whose type is a proposition). The rest of them, we just call variables.

Reply:
How about this?  I think we can still consider variables to be hypotheses--its makes the definition here simpler (only "hypotheses above the line", not "hypotheses and variables above the line").

`Hypotheses that just associate a type with a name, such as `n : nat` are referred to as *variables*.`

Reply:
The distinction is purely informal (irrelevant from the point of view of Coq) but it may still be counter-intuitive to use the same name for variables and hypotheses for mathematicians.

However, your proposal to define "variable" does not work since it applies to all hypotheses. `H` is just a name to which you associate a type (the statement of the hypothesis). The only thing in the context to which it wouldn't apply is local definitions (`n := 0 : nat`).

Reply:
In many places in the text we refer to "hypotheses".  Where the doc uses "hypotheses", do you think that already excludes variables and constants (so we don't have to review/revise all uses of "hypotheses"?

Do we need a term to refer to everything above the line?  I'd rather not have to list the possibilities wherever such a concept is needed.  Neither "facts" nor "above the line" would be good for general use.

I reworded this.  If you still don't like it perhaps you would suggest some wording?

> what we usually call "hypothesis" are only the "H" (i.e. the variables whose type is a proposition)

Meaning `Prop` or some other definition of "proposition"?  The former seems inconsistent with the behavior of the `type of` construct:

```
Goal forall n m : nat, n > m -> True.
intros.
let x := type of H in idtac x.          (* (n > m) *)
let y := type of (n > m) in idtac y.    (* Prop *)
```

Reply:
>  Where the doc uses "hypotheses", do you think that already excludes variables and local definitions

Variables, no (or extremely rarely), because variables are just non-propositional hypotheses and hypotheses are propositional variables (we can choose whichever term we prefer as the "base" one). Local definitions, yes, they are often handled differently.

>Do we need a term to refer to everything above the line?

Doesn't the "proof context" term already encompasses this concept?

> Meaning `Prop` or some other definition of "proposition"?

Yes, "is a proposition" generally means "whose type is `Prop`" (or `SProp`): `H` is a hypothesis because its type `n > m` is a proposition because its type is `Prop`. Thus, it's expected that you need two `type of` to get to `Prop`. By contrast, `n` is a variable, because its type is `nat`, whose type is `Set`.

Reply:
Check what I wrote here and in the revised text in the tactics chapter.

> >Do we need a term to refer to everything above the line?

> Doesn't the "proof context" term already encompasses this concept?

I guess the items above the line should just be called "hypotheses".

I expect "context" would include both the hypotheses and the conclusion.  "proof context" has all goals; the entire proof state.  "goal context" has the hypotheses and conclusion for a single goal.  Does that make sense?

Reply:
It seems to me that the standard definitions of a "foo context" are what is around foo but not foo itself. So "proof context" does not include the entire proof state in my opinion. We can also call it "hypotheses", but it would be less precise, because of the ambiguity with "propositional hypotheses".

Reply:
For "foo context", yes and no.  The point of using "context" is to highlight the relationship between "foo" and things it relates to.  The dictionary says "context" is: "the parts of something written or spoken that immediately precede and follow a word or passage and clarify its meaning" and that "in context" means "considered together with the surrounding words or circumstances."

"Goal context" makes sense (i.e. everything above the line).  "Proof context" doesn't seem as useful/sensible and I think we should drop it.  WDYT?

There are:
- 397 uses of "context" in the doc.  They would need to be reviewed individually to see what they refer to and whether some should be changed.  (Not going to do that in this PR.)
- 25 uses of "proof context", 16 of which are in SSR.  Most of the rest are in text I've previously updated.
- 6 uses of "goal context"
- 6 uses of "Contextual patterns" in SSR but it seems to be a different meaning of "context".  -Perhaps we can find a different term?
- 3 uses of "section context"  is used twice in SSR and one in `type-classes.rst`--maybe that term can be dropped.

Reply:
>"Goal context" makes sense (i.e. everything above the line). "Proof context" doesn't seem as useful/sensible and I think we should drop it. WDYT?

Agreed.